### ISSUE TITLE
[2.2.1 beta] 3D graphics not working properly on Debian XFCE x64 VM

### ISSUE THREAD

**Issue by SugarRayLua** [utmapp/UTM#3060]
UTM 2.2.1 beta fixed memory leak on my custom Debian XFCE x64 VM build and able to render older 3D accelerated graphics programs well using Qemu Display virtio-ramfb; however:

1. Qemu Display virtio-GPU-gl-pci does not work at all (only see black screen on boot of VM)

2. Qemu Display virtio-ramfb does not seem to work with newer accelerated graphics programs (e.g. Linux version of Portal app installed on VM through steam). VM with Qemu Display virtio-ramfb attempts to open Portal app but then stops and displays a small window with window label “Error”. (Debug log attached). Yet, same VM set-up using what I believe is an older non-accelerated Qemu Display driver (Spice XL-VGA) does render Portal app fine (although at way too slow of a speed to make the app playable). 

Helpful if accelerated graphics could work better on Debian x64 build to allow use of wider app availability on x64 system and to potentially use in conjunction with wine and Windows apps which are limited to x64 systems 

Thank you.

**Configuration (required)**
* UTM Version: 2.2.1 beta on custom Debian XFCE x64 VM
* OS Version: 14.7.1
* Device Model: iPad Pro M1 256GB
* Is it jailbroken (name jailbreak used)? No
* How did you install UTM? Signulous

**Debug log**

[debug-1.log](https://github.com/utmapp/UTM/files/7069137/debug-1.log)

**Comment by SugarRayLua** [comment]
OSY, would you provide me some brief instructions on how to test the QEMU version at the github site you directed me, too? I apologize but I can’t figure out how to do that with my novice Linux skills. From what I think I understand from reading the modified QEMU site README,

1. Another developer has hacked QEMU to enable QEMU to do openGL accelerated graphics on his new M1 Apple laptop. 

2. However, it seems that the developer’s instructions for using his hack instructs users to go to Terminal on their Apple laptop and build QEMU from the Terminal app. I don’t understand how therefore I would modify those instructions to build QEMU on my iPad Pro? Or were you suggesting building it on an Apple desktop or laptop with Terminal and then exporting QEMU somehow to my iPad Pro and the UTM app? The other problem it would seem that I would run into is that the developer’s hacked QEMU is being built for Arm64, and I was really interested in being able to use it for AMD64 VM. 

3. It looked like if somehow I could get the modified QEMU you’d like me to test to my iPad Pro, I would then go to the setup screen of my Debian 10 x64 XFCE screen and under the “QEMU” heading, select the “Advanced” option and bypass the standard QEMU configuration, and enter the path to the new modified QEMU you’d like me to test (and presumably leave all the other QEMU settings alone such as the paths to the spice tools?) Unfortunately, I’ve also still not figured out how to use the UTM shared directory feature to be able store the modified QEMU if I got it on my iPad Pro and to be able to tell UTM the path to that shared directory. It seems that all I should need to do is select the “browse” option on the “shared directory” UTM main screen of my VM and choose the host shared directory to link to (which I did). However, I’ve never then been able to find the shared directory files when I am inside any of the VM as a guest. It looked like from your instructions that the host shared directory files might be in my /home/username/documents/public folder of my file structure or perhaps /home/username/public but don’t see the shared files in any of those folders. UTM SE seems to self-create a “public” folder in the app that I can see under my iPad files app, but unfortunately when I put things in that public folder, I can’t see those files as a guest in my VM. I’ve tried using sudo find / -iname nameofthefileIputinthefolder and don’t come up with anything. 

I continue to be very excited to use your app and would be happy to continue to do beta-testing but may need more guidance to do so as a novice Linux user.

Thanks,

Mike

PS, I saw from the modified QEMU developer’s video on the developer’s website that they tested their modified QEMU on their Apple M1 laptop with glmark2 and was able to obtain up near 800 fps on that testing! I installed glmark2 as a comparison on my iPad Pro M1’s custom Debian 10 x64 XFCE UTM 2.2.1 using virtio-ramfb and got around 200 fps (similar to glxgears). Do you think that iPad Pro w/UTM and virtio drivers might sometime get close to the 800 fps acceleration speed or do you think the limitations of the iPad OS prevent reaching that same acceleration speed that the modified QEMU developer got on their Apple M1 laptop?

**Comment by osy** [comment]
Sorry I thought you were on macos. It’s different on iPad because it doesn’t have virtualization. If you have a M1 Mac you can follow those directions or easier one linked from that gist https://github.com/knazarov/homebrew-qemu-virgl

If you don’t have M1 Mac then don’t worry about it. Maybe someone else can test it.

**Comment by SugarRayLua** [comment]
Thanks, OSY, for clarifying. No, I don’t have an M1 Mac. Maybe someone else can test it.

**Comment by SugarRayLua** [comment]
PPS,

Perhaps one last piece to the puzzle. When I query my Debian 10 x64 VM while using QEMU SpiceXL display that is able to open and graphically display the Portal app, it says that is is software rendering OpenGL. Seems like some specific bug in virtio-gl implementation that is preventing only certain 3D hardware accelerated apps from running. Virtio-gl display is able to open up and run glxgears, glmark2, DevilutionX and even linux Google Earth apps fine with 3D acceleration (although none of those are specifically games—perhaps certain 3D games are using specific 3D calls that the other 3D apps aren’t that the current UTM virtio-gl doesn’t support?). Will await to see if others have same problem with 3D games or if future UTM revision solves the problem. Fyi.

**Comment by SugarRayLua** [comment]
OSY, another thought:

Could it be a host 3D driver issue (e.g. virglrender version in UTM)? I created a Ubuntu x64 VM to possibly have more baseline graphics drivers installed. I still had the same problem not being able to run Portal with openGL virtio acceleration as in Debian. I then wondered if it might be that the 3D games uses newer openGL versions as compared to glxgears or other 3D demo programs. In noticed that using virtio-vga (or virtio-ramfb) glx info tells me that the standard Ubuntu 20 install gives me a mesa driver with only openGL version 2.1 capabilities and many 3D games ask for openGL of 3 or higher (not sure what Portal requires). So, I tried updating my Ubuntu Mesa driver to Mesa 21.2.1 (Kisangani-mesa PPA) which strangely did not update my openGL version (still at 2.1) despite the latest Mesa driver supposedly being able to support openGL of 4 or higher. Thus, I wondered if the virglrender version in the host might be limiting what openGL version in the guest I can use. 

According to this blog from Collabora 2 years ago, virglrender 0.8 seemed to be able to support the higher openGL versions and even demonstrated that they could run the Portal app with anth version of virglrender:

https://www.collabora.com/news-and-blog/blog/2019/08/28/virglrenderer-state-of-virtualized-virtual-worlds/

I didn’t know what version of virglrender you were using in UTM but thought if it was below 0.8 maybe you could try upgrading virglrender. 

Fyi, thanks

**Comment by osy** [comment]
We're using 0.8.2, but the OpenGL ES implementation on iOS is very old and doesn't support many features. I think the only hope is that the Metal backend will improve in time and we can switch back to it.

**Comment by conath** [comment]
Possible duplicate of https://github.com/utmapp/UTM/issues/3053 but I don't see the same errors in the Debug log, and it's a different host & guest. So I'm not marking as duplicate.

**Comment by SugarRayLua** [comment]
Thank you both, OSY and Conanth. That would seem to indicate that it is not the host virglrenderer that is limiting things since 0.8.2 is supposed to support openGL 3.0 and higher. Still seems strange though that the Ubuntu guest thinks it can only run OpenGL 2.1 from the virtual GPU despite having both a host virglrender and guest Mesa drivers (21+) that support higher openGL. It is my understanding the Kernel in Ubuntu 20 should support openGL 3.0+ also. I’ll go ahead and write the Collabora blogger to see if he had any tips he used in his VM to get these 3D games working or as to why my guest VM doesn’t think it can support openGL 3.0 and let you know if I figure out anything. Have a good week :-)

**Comment by SugarRayLua** [comment]
(Sorry, Conath for initially misspelling your name).

**Comment by osy** [comment]
So looking more info this, there's two OpenGL profiles supported by mesa. Desktop GL (in `glxinfo` it's listed as `OpenGL version string`) and GL ES (in `glxinfo` it's listed as `OpenGL ES profile version string`). I think the desktop GL is limited to 2.0 but GL ES should be 3.0. Can you double check @SugarRayLua ?

**Comment by ktprograms** [comment]
> So looking more info this, there's two OpenGL profiles supported by mesa. Desktop GL (in `glxinfo` it's listed as `OpenGL version string`) and GL ES (in `glxinfo` it's listed as `OpenGL ES profile version string`). I think the desktop GL is limited to 2.0 but GL ES should be 3.0. Can you double check @SugarRayLua ?

@osy This is what I get (VM using `virtio-ramfb`):

```
$ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa/X.org (0x1af4)
    Device: virgl (0x1010)
    Version: 20.2.6
    Accelerated: yes
    Video memory: 0MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: virgl
OpenGL version string: 2.1 Mesa 20.2.6
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 20.2.6
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
```

**Comment by SugarRayLua** [comment]
Hi, @osy & @conath  & @ktprograms,

This is what I have for my glxinfo -B:
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa/X.org (0x1af4)
    Device: virgl (0x1010)
    Version: 21.2.1
    Accelerated: yes
    Video memory: 0MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: virgl
OpenGL version string: 2.1 Mesa 21.2.1 - kisak-mesa PPA
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 2.0 Mesa 21.2.1 - kisak-mesa PPA
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16

I’m not sure why @ktprograms has a higher max compatible OpenGL ES version then me after I installed the latest mesa drivers in Ubuntu. This is from my iPad Pro M1 256G machine UTM 2.2.1 using custom VM Ubuntu 20 x64. What is your VM build @ktprograms?

My hypothesis from what both @osy was saying about OpenGL on the iOS being old and @conath saying that before the UTM 2.2.1 changeover from metal to OpenGl and further reading was that since iOS as a mobile OS only uses OpenGL ES which has a more restricted OpenGL instructions set than OpenGL proper (or than Metal) that some of the 3D Games (e.g. Portal that I tried or the Kart game that @conath tried) would render when UTM passed graphic instructions onto Metal but that OpenGL ES might not have understood some of them and thus not rendered the images. It seemed that might also explain why the Qemu QXL using software rendering in the host (with presumingly understanding full OpenGL instructions) could render Portal (albeit very slowly) whereas virtio could not when sending instructions to the GPU. I thought that hypothesis may also explain why the diablo port to DevilutionX works so well with UTM—since their website indicates that DevilutionX can be built from even mobile device (e.g. android and consoles) that perhaps the DevilutionX uses only OpenGL ES instructions so works with UTM. Macs with OSX I understand are able to use full OpenGL instructions (although depreciated) so perhaps someone with a Mac running UTM with virtio and 3D acceleration could tell us if full 3D games like Portal or the Kart game can run on their virtual machines? 

I have little technical expertise in this area and don’t know how actually the UTM qemu host passes graphics commands to the iPad or Mac GPU so not sure if my hypothesis is sound but hopefully helping to further improve UTM’s graphics capabilities. If I am way off base, and it is just that I needed to update my OpenGL ES Mesa driver in some manner to get it to work, I’d be happy to try whatever someone else found worked to do that.

Fyi & thanks for the continued effort to make an already great app even greater!

**Comment by ktprograms** [comment]
@SugarRayLua my VM is Ubuntu 20.04 arm64 on a M1 Mac with UTM 2.2.1.

**Comment by SugarRayLua** [comment]
@ktprograms, thanks! Maybe that explains the difference in your openGL ES max capability? If you happened to have any 3D games (e.g. Portal-linux version on Steam or other 3D linux games) you could let us know if they run? You’d need to install custom x64 version of Ubuntu though as most of the 3D games (including on Steam) are x64 builds. It was pretty easy to custom install virtual Ubuntu 20.04 x64 from Ubuntu .iso download from Ubuntu’s website once I got the hand of it. :-)

**Comment by SugarRayLua** [comment]
(Although may not need to test if @conath VM is already on a Mac and doesn’t run 3D games)

**Comment by SugarRayLua** [comment]
@osy, the Collabora blogger got back to me this morning with his thoughts and suggestions:

Dear [SugarRayLua],

On 30/08/2021 00:09, [SugarRayLua] wrote:

I’ve tried Debian 10 and most Recently Ubuntu 20 VM as hosts and have even updated my Ubuntu Mesa drivers to the newest (21+) but have not had success. The UTM app qemu port is using virglrenderer version 0.8.2

Gert’s response:

I would suggest to switch to the latest version 0.9.1, or even the latest git master.

[SugarRayLua}:

and glx info indicates the virtio driver is in place (tested using virtio-vga-gl and virtio-ra fx) and is doing accelerated direct rendering (both virtios set up a ram frame buffer). Strangely, despite having the newest Mesa drivers and virglrenderer version 0.8.2 on the qemu host, glxinfo says that my openGL driver only support openGL 2.1 —not sure why it doesn’t say anything higher.

Gert’s response:

I assume that the IPad only runs OpenGL ES, and for that reason we need certain non-standard extensions that might not be implemented in iOS (we also have to lie a bit about what OpenGL is actually supported).

[SugarRayLua]:

I don’t know if the 3D games like Portal aren’t able to run because I can’t get openGL 3.0+ or better rendering or some other reason.

Gert’s response:

I was under the impression that Portal only needs OpenGL 2.1, in any case you can always try to force  another OpenGL version by using MESA_GL_VERSION_OVERRIDE.

Also if you can run steam from the command line, and then start Portal, you might get some more log output that could give an insight of what is going wrong.

[SugarRayLua]:

.  I don’t think it is an issue with my Portal installation (installed Portal for Linux via Steam) since Portal does run on the same setup using Spice QXL driver (just very very slow to not make it really playable).

Gert’s response:

I'm not really familiar with the Spice QXL driver, but wold this actually mean that the game is running on a software renderer, like llvvmpipe?

I hope that this is of some help to you, and please don't hesitate to ask more questions,


Best,

Gert

**Comment by osy** [comment]
@SugarRayLua sorry about the bot, I've copy pasted your comment below:

------------------------

@osy, the Collabora blogger got back to me this morning with his thoughts and suggestions:

Dear [SugarRayLua],

On 30/08/2021 00:09, [SugarRayLua] wrote:

I’ve tried Debian 10 and most Recently Ubuntu 20 VM as hosts and have even updated my Ubuntu Mesa drivers to the newest (21+) but have not had success. The UTM app qemu port is using virglrenderer version 0.8.2

Gert’s response:

I would suggest to switch to the latest version 0.9.1, or even the latest git master.

[SugarRayLua}:

and glx info indicates the virtio driver is in place (tested using virtio-vga-gl and virtio-ra fx) and is doing accelerated direct rendering (both virtios set up a ram frame buffer). Strangely, despite having the newest Mesa drivers and virglrenderer version 0.8.2 on the qemu host, glxinfo says that my openGL driver only support openGL 2.1 —not sure why it doesn’t say anything higher.

Gert’s response:

I assume that the IPad only runs OpenGL ES, and for that reason we need certain non-standard extensions that might not be implemented in iOS (we also have to lie a bit about what OpenGL is actually supported).

[SugarRayLua]:

I don’t know if the 3D games like Portal aren’t able to run because I can’t get openGL 3.0+ or better rendering or some other reason.

Gert’s response:

I was under the impression that Portal only needs OpenGL 2.1, in any case you can always try to force another OpenGL version by using MESA_GL_VERSION_OVERRIDE.

Also if you can run steam from the command line, and then start Portal, you might get some more log output that could give an insight of what is going wrong.

[SugarRayLua]:

. I don’t think it is an issue with my Portal installation (installed Portal for Linux via Steam) since Portal does run on the same setup using Spice QXL driver (just very very slow to not make it really playable).

Gert’s response:

I'm not really familiar with the Spice QXL driver, but wold this actually mean that the game is running on a software renderer, like llvvmpipe?

I hope that this is of some help to you, and please don't hesitate to ask more questions,

Best,

Gert

**Comment by osy** [comment]
So some comments:

* We're using 0.9.0 and it's close to master (we've diverged a little in order to get macOS/iOS working)
* QXL has no hardware acceleration, so you're using SW OpenGL which will always work but is 100% CPU.
* It's particularly annoying to debug because we have many layers of abstraction: iOS/macOS has a system level OpenGL ES to Metal translator. ANGLE uses the system OpenGL ES libraries as a frontend to its own renderer. ANGLE then produces a OGL/OGLES version number corresponding to the highest API that has all the features it can support. In our case, it is 3.0. Virgl then interfaces with ANGLE (which it sees as just the OGL APIs) and queries features that it supports. Those feature bits are then passed to the guest VM's virgl mesa drivers. The mesa driver then makes it's own interpretation of the OGL API version based on what it sees from virgl. This means it's a pain to debug "why" mesa on the guest VM reports that only 2.1 is supported. It could be any number of reasons spanning from virgl bug to ANGLE bug to Apple compatibility. If Gert knows an easier way to figure out what exactly is missing, I could spend more time trying to fix it.

**Comment by SugarRayLua** [comment]
Ok, thanks for explaining all that @osy. I’ll write to Gert with your comments and see what he says.

I can now also understand why I read that one developer is trying to reverse engineer a Mesa driver specifically for the M1 GPU in their efforts to make a linux distribution to M1.

**Comment by 1640675651** [comment]
Hi, @osy, since now the metal backend of ANGLE is not usable and iOS has its own GLES 3.0 implementation, why not directly use iOS's GLES implementation as the backend of virglrenderer? I wonder what's the point of using ANGLE converting GLES to GLES?

**Comment by osy** [comment]
Because Virgl doesn’t support it

**Comment by SugarRayLua** [comment]
@osy, clarification question:
When you mention ANGLE above are you referring to the MetalANGLE fork of ANGLE? I was just wondering as I read there is such a fork, and if you weren’t using it already if that MetalANGLE fork might work any better.

**Comment by osy** [comment]
Both has the same bug with the memory leak.

**Comment by SugarRayLua** [comment]
:-( . Thanks for clarifying.

**Comment by SugarRayLua** [comment]
Wow, @osy, thanks so much! A very nice holiday gift to the UTM community. Suspect will be in the next UTM release and will look forward to trying it out then. :-)

**Comment by SugarRayLua** [comment]
@osy, did you also test the graphics on v.4.1.3 on UTM SE? I downloaded UTM SE v.4.1.3 and all my previous saved VMs on UTM SE crashed when trying to open them-- even if I set their graphics to non-GL accelerated. I tried creating new VMs in UTM SE v.4.1.3. and they also crashed on loading. I was able to download Debian 11 XFCE from the gallery and get it loaded with graphics acceleration on UTM SE v.4.1.3.  However, when I ran glxgears on it, it rendered the gears for about a couple of seconds and then crashed (same when I retried it). 

I'm trying to get a working solution to enable JIT again for my UTM then will download UTM v.4.1.3 and test that out. Thanks!
